byterun/intern.c

```diff
@@ -686,39 +763,54 @@ CAMLexport value caml_input_value_from_malloc(char * data, intnat ofs)
 /* [len] is a number of bytes */
 CAMLexport value caml_input_value_from_block(char * data, intnat len)
 {
-  uint32_t magic;
-  mlsize_t block_len;
+  struct marshal_header h;
   value obj;
 
   intern_input = (unsigned char *) data;
   intern_src = intern_input;
   intern_input_malloced = 0;
-  magic = read32u();
-  if (magic != Intext_magic_number)
-    caml_failwith("input_value_from_block: bad object");
-  block_len = read32u();
-  if (5*4 + block_len > len)
-    caml_failwith("input_value_from_block: bad block length");
-  obj = input_val_from_block();
+  caml_parse_header("input_value_from_block", &h);
+  if (h.header_len + h.data_len > len)
```

Comment:
I might be wrong here, but it seems to me that a maliciously crafted header might cause the path at https://github.com/ocaml/ocaml/blob/e2d0a13165b31cb06d4a7bdba3847198b32e1bd4/byterun/intern.c#L629 to be followed and cause the values for `h.header_len` and `h.data_len` to overflow when added. This would in turn lead this condition to evaluate to true when the programmer intended false, with possibly bad consequences when `input_val_from_block` is entered.